### PR TITLE
[MISC] Speedup rigid solver using CPU skyline-envelope sparse Cholesky.

### PR DESCRIPTION
**Pull request by duburcqa**
## Description

Adds a skyline-envelope (profile) Cholesky to the rigid Newton constraint solver, exploiting the analytical
sparsity of `H = M + J^T D J` (tree-structured mass + contact coupling). The factor visits only each row's
envelope, so it is exact (skips structural zeros only) while avoiding the dense `O(n_dofs^3)` cost.

What a reviewer should know:
- **`sparse_solve` is now tri-state (`None` = auto).** Auto enables it only on **CPU**, not under MuJoCo
  compatibility, and only when the scene has block structure (>=2 DOF-carrying bodies or >=2 free joints). An
  explicit `True`/`False` overrides; `True` on GPU keeps the prior assembly-level sparsity but warns.
- **The envelope factor and its build-time DOF reorder are CPU-only and disabled under `requires_grad`** - the
  differentiable adjoint solve reuses `nt_H` with natural, dense indexing, so under grad only the pre-existing
  assembly-level sparsity applies. On GPU the dense tiled factorization is faster, so the envelope is not used
  there either (GPU behaviour is unchanged from `main`).
- The sparsity pattern is computed analytically (not by scanning the matrix) and folded into `func_solve_init`,
  so there is no extra per-step kernel launch.

## Motivation and Context

On `main`, `sparse_solve` only sparsified assembly while the factor stayed dense, making it a net *slowdown*.
This makes the factor itself sparse. CPU, whole-sim average ms/step (n/a = not yet measured):

| Scene (CPU) | dense | `main` `sparse_solve` | this PR | speedup (dense / this PR) |
|---|---|---|---|---|
| 100-box tower + duck (606 DOF) | 69.0 | 119.8 | **15.9** | **4.3x** |
| shadow hand + 25 cubes (198 DOF) | 20.8 | n/a | **8.5** | **2.4x** |
| 2 robots (70 DOF) | 0.143 | n/a | 0.154 | 0.93x |
| g1 humanoid (35 DOF, single tree) | 0.105 | n/a | 0.108 | 0.97x |

The `this PR` column shows the envelope forced on. The auto-heuristic turns it on where the band is much tighter
than n_dofs (tower, shadow); it stays roughly break-even on small densely-coupled scenes like 2 robots (~8%) and
is left off for a single dense-coupled tree like g1 (so its ~3% overhead is not incurred by default).

## How Has This Been / Can This Be Tested?

- `tests/test_rigid_physics_sparse.py` passes on CPU and GPU.
- `tests/test_grad.py` passes: differentiable multi-entity scenes now auto-enable sparsity, exercising the
  adjoint/envelope interaction (gradients match the dense path).
- Sparse trajectories match the dense reference within FP noise, including scrambled body-insertion order.

## Checklist:
- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [x] I updated the documentation accordingly or no change is needed.
- [x] I tested my changes and added instructions on how to test it for reviewers.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: genesis/engine/solvers/rigid/rigid_solver.py [hunk 1]
```diff
@@ -423,7 +444,8 @@ def _build_static_config(self):
             enable_collision=self._enable_collision,
             enable_joint_limit=self._enable_joint_limit,
             box_box_detection=self._box_box_detection,
-            sparse_solve=self._options.sparse_solve,
+            sparse_solve=sparse_solve,
+            reorder_dofs=sparse_solve and not self._enable_mujoco_compatibility,
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply genesis/engine/solvers/rigid/rigid_solver.py]: fixed.

**Comment by duburcqa** [comment]
@codex review

File: genesis/engine/solvers/rigid/constraint/solver.py [hunk 2]
```diff
@@ -138,6 +139,18 @@ def __init__(self, rigid_solver: "RigidSolver"):
         # and not used when hibernation is not enabled.
         self.contact_island = ContactIsland(self._collider)
 
+        # Fill-reducing DOF permutation for the skyline Cholesky: a structural choice fixed once from the initial
+        # body layout (forward kinematics has already run at this point), never recomputed in the step loop. The
+        # envelope factorization is CPU-only, so the permutation is too.
+        if self.sparse_solve and gs.backend == gs.cpu:
+            func_compute_dof_perm(
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply genesis/engine/solvers/rigid/constraint/solver.py]: Fixed: the DOF permutation is now computed whenever `sparse_solve` is set (both backends), not just CPU. On GPU it resolves to the identity permutation (the COM reorder is gated on the CPU-only envelope), so the sparse Hessian assembly's `dof_iperm` indexing is always initialized.

File: genesis/options/solvers.py [hunk 3]
```diff
@@ -508,7 +513,7 @@ class RigidOptions(Options):
     noslip_iterations: NonNegativeInt = 0
     noslip_tolerance: PositiveFloat = 1e-6
     contact_pruning_tolerance: PositiveFloat | None = 0.02
-    sparse_solve: StrictBool = False
+    sparse_solve: StrictBool | None = None
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply genesis/options/solvers.py:516]: `not None` is `True` in Python
